### PR TITLE
Implement rotation

### LINKED ISSUE #50: Can you support devices rotate ?
[Warning:ButtPlugin] Websocket error: You must send data by websocket after websocket is opened!
[Message:ButtPlugin] Warning: No devices connected to Buttplug.
[Warning:ButtPlugin] Buttplug error: {"Error":{"Id":1622452820,"ErrorCode":4,"ErrorMessage":"{\"ButtplugDeviceError\":\"DeviceScanningAlreadyStarted\"}"}}
[Message:ButtPlugin] Warning: device "Vorze A10 Cyclone SA" not supported.

I hope you can support device rotate :v

### PR DESCRIPTION
**Pull request by hogefugamoga**
close: https://github.com/Sauceke/LoveMachine/issues/50

Hello, thank you for creating awesome plugin.

I have Vorze A10 Cyclone SA and tested this implementation with IO.
It works well and I think I can create `XxxButtplugRotatorController` as well for another games which I don't have and can't test, but I'd like to ask your review as a first step.

### CODE REVIEW

**Comment by Sauceke** [comment]
Hi, first of all don't forget to include yourself [here](https://github.com/hogefugamoga/LoveMachine/blob/rotation/doc/README.md) and [here](https://github.com/hogefugamoga/LoveMachine/blob/rotation/Installer/Readme.rtf) as a contributor.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [inline LoveMachine.Core/ButtplugController.cs]: You should probably include clockwise as a parameter here. Then you can demote both random and clockwise to local variables in RunRotatorLoop.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [inline LoveMachine.Core/ButtplugController.cs]: Why aren't we waiting for the up-stroke after this line?

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [inline LoveMachine.Core/ButtplugController.cs]: This will always be false since there is no yield between the previous line and this.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [inline LoveMachine.Core/ButtplugController.cs]: I might be missing something here; won't this just rotate in one direction at a steady pace for the full duration of the stroke? Is there any variation besides occasionally changing direction?

Comment (Sauceke) [inline LoveMachine.Core/ButtplugController.cs]: I normally try to avoid using named arguments when they already have the same name as the parameter because it feels redundant.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [inline LoveMachine.Core/ButtplugController.cs]: Okay but why stop the rotation for just one frame?

**Comment by Sauceke** [comment]
What is the effect you're trying to achieve here? I don't know much about rotators, but I can see 2 different ways of using them:
- stroker-like (rotate one way at constant speed during up-stroke and the other way during down-stroke)
- vibrator-like (rotate in one direction but at a changing speed, gradually increase speed during down-stroke and decrease speed during up-stroke)

It seems to me that you're trying to do something like the first one. Did you consider the second option as well? You can implement it quite easily using the vibrator loop as a reference, although I'm not sure how many commands per second the Cyclone can take.

**Comment by ghost** [comment]
@Sauceke I don't have an exact quantity, but I expect my UFO SA to be unable to keep up with being treated as a vibrator. I can research this if you make a vibrator-like rotator implementation (by probably just messily substituting in RotateCmds), although it may not be worth the effort.
I don't think the updates can occur quick enough to have smooth oscillations, hence the stroker-like approach you cited is probably better.

I didn't read the code, so I don't know what @hogefugamoga is going for. I'll test stuff once you guys have touched base.

**Comment by Sauceke** [comment]
@YuriseXYZ Intiface Desktop has a UI in the Devices section where you can test toys in real time. I don't know what it looks like for rotators, but I assume it has a slider for the speed that you can drag around and see if the transition is smooth enough.

Also, how are rotators used in other projects? Like what does a typical funscript for rotators do? If none of them uses the vibrator style approach, it's probably because it's either unfeasible or unenjoyable.

**Comment by ghost** [comment]
@Sauceke I'm not in a spot where I can check right now. I'll get back to you on that. I only have a UFO SA so I can't speak for the cyclone. The scripts that I've encountered for rotators are the vorze .csv style ones. If I remember correctly, rotation speed and direction are synced to whatever media on the scale of seconds, rather than frames (I forgot what scale funscripts are on, frames?). The scripts that I've encountered rely on a stroker style approach, as it's probably a pain to manually script oscillations and ramps. The action would probably happen too quickly for a vibrator style approach to work when scripting anyway.

The basis of my claim that rotators can't handle vibrator-levels of message frequency is on my attempt at modding [music-vibes](https://github.com/Shadlock0133/music-vibes) with rotation instead of vibration by just dropping in RotateCmds where VibrateCmds are. I don't know Rust or audio stuff so I can't tell what frequency the devices are updated at by reading the source.

I am not aware of any actual project that uses rotators aside from イクノガマン (a proprietary game using Buttplug Unity) and some script players. I don't recall it being synced to the fingers in that game, but I would call their approach closer to the stroker style with constant speeds and changing directions. Apparently some eroge exist for use with the cyclone X10, but I have no idea what they do there.

The vibrator approach doesn't really give as strong as an impression that the device is synced to movements compared to the stroker style approach. I don't know what things will look like for either approach when there is fast motion on the screen though.

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: Yeah, I did such way but I noticed we need to reimplement `clockwise` handling in `TestRotator` as well.
I this it's better to encapsulate this logic in DoRotate since currently it's just changed randomlly.

```
if (random.NextDouble() <= CoreConfig.RotationDirectionChangeChance.Value)
            {
                clockwise = !clockwise;
            }
```

What do you think?

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: Hmm, I don't think a rotator needs to wait for that since there's no top side

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: If animation keeps the same speed, there's no variation about speed here.

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: For some reason, rotators (at least A10 cyclone) can't handle same-ish timing command well.
Let's say we have the following code:

```cs
client.RotateCmd(speed: 0, clockwise: clockwise, girlIndex, bone);
client.RotateCmd(speed: 1, clockwise: clockwise, girlIndex, bone);
```

We expect it rotates in full speed but actually it randomly stops forever.

However it seems having sleep works anytime:

```cs
client.RotateCmd(speed: 0, clockwise: clockwise, girlIndex, bone);
yield return new WaitForSecondsRealtime(0.01f);
client.RotateCmd(speed: 1, clockwise: clockwise, girlIndex, bone);
```

I guess rotators confuse commands order if multiple commands are sent in the same time.

That's why we need sleep here.

**Comment by hogefugamoga** [comment]
@Sauceke 

> What is the effect you're trying to achieve here? I don't know much about rotators, but I can see 2 different ways of using them:

I thought stroker-like is better but as you mentioned vibrator-like implementation is not difficult.
I'll send another PR to achieve that but I have a question.

*How do we decide which one is better?*

Should I provide option to choose pattern and wait for users' feedback?

**Comment by ghost** [comment]
@hogefugamoga I think that stroker-like is better too. I don't think we need to do a second vibrator option.

How does it feel with a fast animation though?

**Comment by Sauceke** [comment]
@YuriseXYZ [These are the artifacts from the PR build](https://github.com/Sauceke/LoveMachine/suites/6449904222/artifacts/236669279). If you have time, could you please tell me if it does what you'd expect (and if it doesn't, what should be changed)? I tried to test it but I can't create virtual devices anymore for some reason, Intiface keeps throwing a weird error.
It's only enabled in Insult Order, if you don't have that game, I can quickly turn it on for other games with just a couple of lines.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [reply LoveMachine.Core/ButtplugController.cs]: I mean why stop it in the first place if it's only for one frame? Why not just leave it on? What difference does it make if it stops for one frame or keeps moving for one frame?

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [reply LoveMachine.Core/ButtplugController.cs]: I see, in that case my question is: What is this movement synced to? Isn't it basically just doing its own thing with little to no regard to what's happening on the screen? And if so, what is the benefit of having this logic inside the plugin and not in a separate script?

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [reply LoveMachine.Core/ButtplugController.cs]: I don't really get why randomness is needed for testing. Why not just alternate back and forth?

**Comment by Sauceke** [comment]
@hogefugamoga Given that we have 2 people here who actually own a rotator and they both think stroker-like is better, let's focus on that.

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: > why randomness is needed for testing

I think it's better to adjust `CoreConfig.RotationDirectionChangeChance.Value` via test button though, if you think it isn't needed, I'll apply changes

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: > What is this movement synced to?

The movement synced to the speed of strokes.

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: I see. I just copied and pasted stroker's one 🙈

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: > Why not just leave it on?

Ah, I understand what you meant.
I thought it's better to clean up (stop) rotator here but actually in the rotator loop it'll be stopped `OnEndH`, not here.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [reply LoveMachine.Core/ButtplugController.cs]: I don't want to micromanage things here, but I would imagine there is so much more that could be done with a rotator. Maybe rotate fast when the pp is going in, and a bit slower when it's going out? So that it feels like something is pulsating or idunno. That would be just two messages per stroke, any rotator should be able to handle that much.

If the current logic provides better experience for you, that can be done in much less code. Just check the stroke time every second or so and set the speed to its inverse. No need for most of the code copied over from the stroker loop.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [reply LoveMachine.Core/ButtplugController.cs]: You don't have to worry about that, because device testing is just an ugly hack at this point, even for strokers. So I'll have to fix that eventually anyway.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [reply LoveMachine.Core/ButtplugController.cs]: Nope, OnEndH is called when exiting H. It should be stopped when IsIdle is true.

**Comment by ghost** [comment]
@Sauceke Yeah I've got Insult Order. I'll test some time soon.

**Comment by ghost** [comment]
@Sauceke Testing is done.

For reference, this is what I think the stroker-like approach is:
>stroker-like (rotate one way at constant speed during up-stroke and the other way during down-stroke)

Of course, you could do stuff like making speeds different within the up-down set as you've suggested (or leave it to `IsHardSex()`), but that's the fundamental idea.

For what I tested, the rotators don't change directions within the up-down set (as in down, change direction, up). Instead, it has the probability of changing directions after a whole up-down stroke 'set' (down, up, then maybe changing directions). Otherwise, the speeds change with faster animation speeds as expected.

I actually don't mind the directions changing after the stroke set, but I want to test a version that faithfully implements the stroker-like approach to see if changing directions within the stroke feels better. In theory, it sounds more haptic, but it might just result in the rotator just jerking about for fast animations and not moving very far.

On the topic of testing rotators in Intiface, the speed changes after you finish dragging the slider (letting go of the mouse), rather than whilst you're dragging it, so that may be a sign that it can't handle rapid device messaging.

**Comment by Sauceke** [comment]
@YuriseXYZ Thank you for confirming and opining on the behavior.
@hogefugamoga Would you be happy to experiment with rotation patterns in this vein? Here are two I think might be worth investigating:
1. always rotate CW but rotate faster on the down-stroke (maybe randomly change direction every once in a while)
2. always rotate with the same speed, but CW on the down-stroke and CCW on the up-stroke

Both can be derived from the stroker loop.

**Comment by Sauceke** [comment]
@hogefugamoga, do you still want to work on this PR? Would you like me to take over from here? I'm concerned that your work might get out of date with the base branch.

**Comment by hogefugamoga** [comment]
@Sauceke Sorry for being late.
I applied some fixes. Can you review again?

File: README.md
Comment (Sauceke) [inline README.md]: This file gets overwritten after each release, please edit the doc/README.md file instead.

**Comment by Sauceke** [comment]
Could you please also edit doc/README.md to reflect these changes ("This plugin is for linear (moving back-and-forth), vibrating **and rotating** sex toys", etc.).

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [inline LoveMachine.Core/ButtplugController.cs:236]: 1. In your experience, is this behavior better than the previous one?
2. Have you tried what it's like when you don't change direction here but keep everything else the same (only remove the "!")?

**Comment by Sauceke** [comment]
@YuriseXYZ Now it rotates back and forth and also changes speed. [Here are the artifacts](https://github.com/Sauceke/LoveMachine/suites/6798036738/artifacts/260988214). I'd love to hear your ~output~ input on this, whenever you feel like testing it.

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs:236]: > In your experience, is this behavior better than the previous one?
> Have you tried what it's like when you don't change direction here but keep everything else the same (only remove the "!")?

I tried both but TBH it's sameish for me.

I implemented both of them as you comment.
https://github.com/Sauceke/LoveMachine/pull/97#issuecomment-1124184472

Comment (Sauceke) [reply LoveMachine.Core/ButtplugController.cs:236]: Then perhaps this could benefit from a setting for whether to change direction here?
`clockwise ^ CoreConfig.RotateBackwardsOnDownStroke.Value` instead of `!clockwise`, perhaps?
Just a suggestion though, it's your call.

File: LoveMachine.KK/KoikatsuButtplugController.cs
Comment (Sauceke) [inline LoveMachine.KK/KoikatsuButtplugController.cs]: You could get rid of lines 246-266 by inheriting from KoikatsuButtplugStrokerController (or by having both inherit from a common base).

File: LoveMachine.KK/KoikatsuButtplugController.cs
Comment (Sauceke) [inline LoveMachine.KK/KoikatsuButtplugController.cs]: This is mainly for female players and IMHO the only sensible way to implement it with rotators would be to set the rotator speed to `y` (which is a value between 0 and 1). Or if you don't want to implement it, you can just throw `NotImplementedException` here and remove the aibu rotator controller.
(It would absolutely make sense to implement it though, as some female toys have rotation, like the Nora).

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: @Sauceke I've noticed sometime rotator doesn't stop after H.
I've confirmed this behavior on KKS.
It happens once about every three or four times.
I guess sometimes OnEndH and/or IsIdle doesn't work well for vibrator as well.
Do you have any idea?

File: LoveMachine.KK/KoikatsuButtplugController.cs
Comment (Sauceke) [inline LoveMachine.KK/KoikatsuButtplugController.cs]:
```suggestion
    internal class KoikatsuButtplugRotatorController : KoikatsuButtplugStrokerController
```

File: LoveMachine.KK/KoikatsuButtplugController.cs
Comment (Sauceke) [inline LoveMachine.KK/KoikatsuButtplugController.cs]: IsIdle, supportedAnimations and supportedModes can be deleted

File: LoveMachine.KK/KoikatsuButtplugController.cs
Comment (Sauceke) [inline LoveMachine.KK/KoikatsuButtplugController.cs]: This has to be deleted.

**Review by Sauceke — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
@Sauceke I tested it on Insult Order and KKS and it's pretty much what I wanted, plus the adjustable randomness which I also liked. I didn't encounter any game-breaking bugs.

> Now it rotates back and forth **and also changes speed**

The only instances of speed changing that I noticed were in the use of slider in the config and in accordance with the animations. I didn't notice changes of speed anywhere else. If that was intended, then everything is as expected.

Other than that, good job to the two of you.

**Comment by Sauceke** [comment]
@YuriseXYZ, according to the code, it should rotate 25% faster in one direction than the other. Perhaps the UFO has too few speed levels and Buttplug rounded them to the same value, or maybe the change is just too subtle... At any rate, thank you again for testing.

**Comment by Sauceke** [comment]
@hogefugamoga, thank you for putting up with my requests. Here are all the remaining changes I'm going to need before I can merge this:
1. The previous 3 suggested changes
2. Add rotation controllers to all games (you don't need to test them, it's all boilerplate anyway)
3. Document the rotation settings in doc/README.md

**Comment by hogefugamoga** [comment]
@Sauceke What do you think?

https://github.com/Sauceke/LoveMachine/pull/97#discussion_r891205169

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs:236]: After several try, I feel `!clockwise` at pull back is important.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [reply LoveMachine.Core/ButtplugController.cs]: Oh yes, it actually has to be stopped in both places. Sorry for the confusion.
When an H scene ends, all coroutines are immediately stopped, so we have to clean up after them, including resetting all device states. This is what OnEndH() does.
So if you add a loop for stopping rotators right below the one that stops vibrators, it should solve the problem.

**Comment by Sauceke** [comment]
@hogefugamoga Sorry, didn't notice that comment. Replied.

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [inline LoveMachine.Core/ButtplugController.cs]: @Sauceke Sorry, I'm not familiar with C# though, is this implementation not enough to stop rotators?

https://github.com/Sauceke/LoveMachine/pull/97#discussion_r891205169
https://github.com/Sauceke/LoveMachine/pull/97#discussion_r895216141

I had added this logic but it still doesn't stop sometimes.

**Comment by hogefugamoga** [comment]
> @hogefugamoga, thank you for putting up with my requests. Here are all the remaining changes I'm going to need before I can merge this:
> 
> 1. The previous 3 suggested changes
> 2. Add rotation controllers to all games (you don't need to test them, it's all boilerplate anyway)
> 3. Document the rotation settings in doc/README.md

Done.

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [reply LoveMachine.Core/ButtplugController.cs]: This will start a coroutine in the same frame StopAllCoroutines() is called. I'm not sure how Unity handles that.
Have you tried calling the client instead?
```suggestion
                    client.RotateCmd(0f, clockwise, girlIndex, bone);
```

File: LoveMachine.Core/Buttplug/ButtplugWsClient.cs
Comment (Sauceke) [inline LoveMachine.Core/Buttplug/ButtplugWsClient.cs]:
```suggestion
        public void RotateCmd(float speed, bool clockwise, int girlIndex, Bone bone)
```

File: LoveMachine.Core/ButtplugController.cs
Comment (Sauceke) [inline LoveMachine.Core/ButtplugController.cs]: I think it's bad practice to treat 0 as a special value. Instead, please either call client.RotateCmd directly when the rotator needs to be stopped, or make a StopRotator method.

File: LoveMachine.Core/PluginInitializer.cs
Comment (Sauceke) [inline LoveMachine.Core/PluginInitializer.cs:346]: I just noticed this now, but since you're adding a new column, `columns` should be increased to 7.
Sorry, IMGUI doesn't have flexible layout capabilities, it's done manually.

**Review by Sauceke — CHANGES_REQUESTED** [review]

File: LoveMachine.Core/ButtplugController.cs
Comment (hogefugamoga) [reply LoveMachine.Core/ButtplugController.cs]: @Sauceke Thanks, the code works well.

**Comment by hogefugamoga** [comment]
@Sauceke Done.

**Review by Sauceke — APPROVED** [review]
Looks good to me. Thank you for your contribution!

**Pull request merged**
